src/OpenLoco/src/Vehicles/Vehicle.cpp

```diff
@@ -116,14 +130,331 @@ namespace OpenLoco::Vehicles
+                if (distance >= 12)
+                {
+                    continue;
+                }
+
+                const auto subType = vehicleBase->getSubType();
+                // Does it actually have a collidable body
+                if (subType != VehicleEntityType::body_continued && subType != VehicleEntityType::body_start && subType != VehicleEntityType::bogie)
+                {
+                    continue;
+                }
+
+                if (vehicleBase->owner != bogie.owner)
+                {
+                    continue;
+                }
+
+                // This is an optimisation compared to vanilla
+                if (vehicleBase->getHead() != bogie.head)
+                {
+                    return vehicleBase->id;
+                }
+
+                bool noSelfCollision = false;
+                {
+                    Vehicle collideTrain(vehicleBase->getHead());
+                    bool carFound = false;
+                    uint32_t numCarsToCheck = 0;
+                    for (auto& car : collideTrain.cars)
+                    {
+                        for (auto& carComponent : car)
+                        {
+
+                            if (!carFound)
+                            {
+                                carComponent.applyToComponents([&carFound, &numCarsToCheck, targetId = vehicleBase->id](auto& c) {
+                                    if (c.id == targetId)
+                                    {
+                                        carFound = true;
+                                        numCarsToCheck = 3;
+                                    }
+                                });
+                            }
+                            if (carFound)
+                            {
+                                carComponent.applyToComponents([&noSelfCollision, targetId = bogie.id](auto& c) {
+                                    if (c.id == targetId)
+                                    {
+                                        noSelfCollision = true;
+                                    }
+                                });
+                            }
+                        }
+                        if (noSelfCollision)
+                        {
+                            break;
+                        }
+                        if (carFound)
+                        {
+                            numCarsToCheck--;
+                            if (numCarsToCheck == 0)
+                            {
+                                break;
+                            }
+                        }
+                    }
+                }
+                if (noSelfCollision)
+                {
+                    continue;
+                }
+                {
+                    bool carFound = false;
+                    uint32_t numCarsToCheck = 0;
+                    for (auto& car : srcTrain.cars)
```

Comment:
todo deduplicate

Reply:
Is this a todo for this PR?

Reply:
should probably do it in this pr. Do you think i should change it as well so that it only checks the next car. I think CS made it do a number of car components rather than mess about counting cars but we have easy access to iterating by car so i made it more like how i thought it should work.